### PR TITLE
[WIP] add tests: sp_auth

### PR DESCRIPTION
**Pull request by nsakaimbo**
### Description

Exploration of how to unit-test the external APIs of the various  modules under `apps`. Uses `meck` to mock side-effect interactions (e.g. database calls). Perhaps this may not be the route we ultimately go but there are some useful/interesting learnings here nonetheless. 

### To Test

```
> rebar3 compile
> rebar3 eunit suite=sp_auth
```

### CODE REVIEW

File: apps/sp_auth/src/sp_auth_tests.erl [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            ?assertEqual(ExpectedAppId, AppId ),
+            ?assertEqual(ExpectedCollectionName, Collection),
+            #{
+                <<"_id">>       => <<"test-api-key">>,
+                <<"admin">>     => false,
+                <<"read_only">> => false,
+                <<"auth_only">> => false
+            }
+        end),
+        { Status, Any } = sp_auth:authorize_request(ExpectedAppId, <<"test-api-key">>),
```
Comment (nsakaimbo) [inline apps/sp_auth/src/sp_auth_tests.erl:24]: **LOL**. The Erlang is weak here. 😅  Wanted to assert like `?assertEqual( { granted, _ }, ... )` (i.e where we don't care for the following parameter), but couldn't figure out how to do so.

Comment (dmsnell) [reply apps/sp_auth/src/sp_auth_tests.erl:24]:
```erlang
?assertMatch({granted, _}, sp_auth:authorize_request(…)),
```

**Comment by dmsnell** [comment]
You picked a hard one to start with 😄 

I'm torn on the general approach. I do feel like we are going to have a tension between time spent making the app work plus adding the integration tests we'll need _for sure_ while also _recreating external service logic_ in tests via mocks and then testing those mocks.

Obviously the database is an obvious candidate for mocking. I'm a bit worried however about the limitations we have in this avenue. Once we go one or more levels deeper into the logic we start to have much more inter-related database activity. The way most of the functions are written, we _can_ test that each stage in the logic chain does what we expect, but at the same time once we start one of those chains it carries to the end. I've considered rearranging the code so we can test each step but I don't know if it's worth writing in a more awkward and non-standard style everywhere just so that we can unit-test the things that demand integration suites.

**Comment by dmsnell** [comment]
@nsakaimbo I've spent more time thinking about this test and I want to encourage the work you're doing; at the same time I think there's a better opportunity for us to bite the bullet and establish our use of `common_test` in the project. We're going to need integration/e2e tests anyway and this particular test really is that, even if we mock it out.

How would you like to try setting that up?

Either way I don't think mocking is going to be the most valuable thing for us. If we want to fake the database responses we should be able to circumvent mocking by calling the appropriate functions with the database responses as the last parameter, as most functions are written in this way that lets us do that and if we have one that isn't, maybe we should double-check that we don't want to rewrite it.

One caveat is that we might be one small step away from making the integration tests viable: we should be able to create and delete a new "app" from the Erlang code so that we can run the tests on clean data and then clean up when we're done. We can look into that work in another PR or in a pairing session if you are interested.

**Comment by nsakaimbo** [comment]
@dmsnell This was a fun/useful exercise in figuring out how to get unit test mocks set up in Erlang, but I do agree that a good suite of integration tests would give the most bang for our buck.

> How would you like to try setting that up?

Not sure yet 😅 - I've been in a Calypso headspace the last week, but intend on getting back to Erlang things over the next couple days. I'll be researching `common_test` things, but to your point it does seem like the next reasonable step would be to figure out how to spawn a new "app" instance for testing purposes. Will pick your brain on that one.

**Pull request closed without merging**